soc: arm: atmel_sam: Clock clean-up

nandojve opened the pull request:
Small clock clean-up. This moves all devicetree entries on SoC to atmel_sam_dt.h.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

stephanosio commented on soc/arm/atmel_sam/common/atmel_sam_dt.h:
```diff
@@ -43,4 +45,17 @@
 	}
 
+/* Devicetree related macros related to clock */
+#ifndef CONFIG_SOC_ATMEL_SAM_MDIV
+#define CONFIG_SOC_ATMEL_SAM_MDIV 1
+#endif
```
If we are going to refactor clock, we should do it properly too (i.e. clock generator node in DT + each device referencing the node with proper bus/div params).

For example, see NXP (SIM) and ST (RCC) device tree:
https://github.com/zephyrproject-rtos/zephyr/blob/e7c7f911a9cd99768c9051b30c2f206e9998de1c/dts/arm/nxp/nxp_k6x.dtsi#L135-L141

https://github.com/zephyrproject-rtos/zephyr/blob/e7c7f911a9cd99768c9051b30c2f206e9998de1c/dts/arm/nxp/nxp_k6x.dtsi#L169

In case of Atmel SAM series, this would be called `ckgr`.

nandojve replied:
Hi @stephanosio, the idea here is only do a simple clean up and moving from SoC about all DT_ related to new file atmel_sam_dt.h. I don't have any intention to touch on clocks right now.
This is the only one and now we can assume on Atmel SAM SoC all DT related things have this file as entry point.

galak commented on soc/arm/atmel_sam/common/atmel_sam_dt.h:
```diff
@@ -43,4 +45,17 @@
 
+/* Devicetree related macros related to clock */
+#ifndef CONFIG_SOC_ATMEL_SAM_MDIV
+#define CONFIG_SOC_ATMEL_SAM_MDIV 1
```
Let's not define CONFIG_SOC_ATMEL_SAM_MDIV, just change the code below to be:


```
#ifdef CONFIG_SOC_ATMEL_SAM_MDIV
#define SOC_ATMEL_SAM_MCK_FREQ_HZ \
		(SOC_ATMEL_SAM_HCLK_FREQ_HZ / CONFIG_SOC_ATMEL_SAM_MDIV)
#else
#define SOC_ATMEL_SAM_MCK_FREQ_HZ SOC_ATMEL_SAM_HCLK_FREQ_HZ
#endif

nandojve replied:
In fact, I changed the name because SAME70 is CONFIG_SOC_ATMEL_SAME70_MDIV, SAMV71 is CONFIG_SOC_ATMEL_SAMV71_MDIV and SAM4xxx don't have it. I tried to keep it simple but can rollback, not big deal.

nandojve committed:
soc: arm: atmel_sam: Clock clean-up

Small clock clean-up. This moves all devicetree entries on SoC to
atmel_sam_dt.h.

Signed-off-by: Gerson Fernando Budke <nandojve@gmail.com>

stephanosio commented on soc/arm/atmel_sam/common/atmel_sam_dt.h:
```diff
@@ -43,4 +45,21 @@
+
+/** Processor Clock (HCLK) Frequency */
+#define SOC_ATMEL_SAM_HCLK_FREQ_HZ \
+		DT_PROP(DT_NODELABEL(cpu0), clock_frequency)
```
We might as well do `DT_INST_PROP(0, clock_frequency)` here.

stephanosio commented on soc/arm/atmel_sam/common/atmel_sam_dt.h:
```diff
@@ -43,4 +45,21 @@
+		DT_PROP(DT_NODELABEL(cpu0), clock_frequency)
+
+/** Master Clock (MCK) Frequency */
+#if defined(CONFIG_SOC_ATMEL_SAME70_MDIV)
+#define SOC_ATMEL_SAM_MCK_FREQ_HZ \
+		(SOC_ATMEL_SAM_HCLK_FREQ_HZ / CONFIG_SOC_ATMEL_SAME70_MDIV)
+#elif defined(CONFIG_SOC_ATMEL_SAMV71_MDIV)
+#define SOC_ATMEL_SAM_MCK_FREQ_HZ \
+		(SOC_ATMEL_SAM_HCLK_FREQ_HZ / CONFIG_SOC_ATMEL_SAMV71_MDIV)
+#else
+#define SOC_ATMEL_SAM_MCK_FREQ_HZ SOC_ATMEL_SAM_HCLK_FREQ_HZ
+#endif /* Master Clock (MCK) Frequency */
```
This is not really a device tree-derived value; I am not sure if it belongs in the `atmel_sam_dt.h`.

nandojve replied:
agreed

stephanosio reviewed (commented):
In general, I am not sure about moving `SOC_ATMEL_SAM_HCLK_FREQ_HZ` and `SOC_ATMEL_SAM_MCK_FREQ_HZ` to `atmel_sam_dt.h` since these are not entirely derived from the device tree.

For the time being, I believe it is more advantageous to keep them in the each `soc.h`, as that is more straight forward and we do not need extra `ifdef`s that way.

These macros should be moved to `atmel_sam_dt.h` when we do proper device tree-based clock system specification, [as I previously commented](https://github.com/zephyrproject-rtos/zephyr/pull/24487#discussion_r410816280).

stephanosio commented on soc/arm/atmel_sam/same70/soc.h:
```diff
@@ -67,8 +67,11 @@
 
-/* Add include for DTS generated information */
-#include <devicetree.h>
+/** Processor Clock (HCLK) Frequency */
+#define SOC_ATMEL_SAM_MCK_FREQ_HZ \
+		(ATMEL_SAM_DT_CPU_CLK_FREQ_HZ / CONFIG_SOC_ATMEL_SAME70_MDIV)
```
This does not seem right:
```
#define SOC_ATMEL_SAM_HCLK_FREQ_HZ ATMEL_SAM_DT_CPU_CLK_FREQ_HZ
```

nandojve replied:
Fixed!

stephanosio commented on soc/arm/atmel_sam/samv71/soc.h:
```diff
@@ -68,8 +68,11 @@
 
-/* Add include for DTS generated information */
-#include <devicetree.h>
+/** Processor Clock (HCLK) Frequency */
+#define SOC_ATMEL_SAM_MCK_FREQ_HZ \
+		(ATMEL_SAM_DT_CPU_CLK_FREQ_HZ / CONFIG_SOC_ATMEL_SAMV71_MDIV)
```
same here

nandojve replied:
Fixed!

nandojve commented:
rebase

stephanosio commented on soc/arm/atmel_sam/sam3x/soc.h:
```diff
@@ -35,18 +35,15 @@
-
-#define ID_UART0   ID_UART
-#define UART0      UART
+#include "soc_pinmap.h"
```
Is this necessary?

nandojve replied:
No more, I remember now.

ioannisg commented:
@galak pls, see if your fb is addressed

nandojve commented:
rebase and switch from DT_NODELABAL to DT_PATH following same approach on #24602.
